### ISSUE TITLE
Unsafe impl Send/Sync allows for memory-unsafe usage

### ISSUE THREAD

**Issue by Ten0** [Robbepop/string-interner#24]
I love the idea of the crate, and I was seriously considering using it, but I feel like it's really overusing `unsafe`, leading to memory bugs. :(

POC is a bit longer than in #19.

```rust
use {
	std::{hash::BuildHasher, rc::Rc},
	string_interner::{backend::Backend, DefaultHashBuilder, StringInterner, Symbol},
};

#[derive(Clone)]
struct RcHasher<H: BuildHasher> {
	inner: Rc<H>,
}

impl<H: BuildHasher> BuildHasher for RcHasher<H> {
	type Hasher = H::Hasher;
	fn build_hasher(&self) -> Self::Hasher {
		self.inner.build_hasher()
	}
}

#[derive(Default)]
struct Stub;
impl<S: Symbol> Backend<S> for Stub {
	fn with_capacity(_cap: usize) -> Self {
		Self
	}
	fn intern(&mut self, _string: &str) -> S {
		unimplemented!()
	}
	fn resolve(&self, _symbol: S) -> Option<&str> {
		unimplemented!()
	}
	unsafe fn resolve_unchecked(&self, _symbol: S) -> &str {
		unimplemented!()
	}
}

fn main() {
	let hasher = RcHasher {
		inner: Rc::new(DefaultHashBuilder::default()),
	};
	let mut threads = Vec::new();
	for _ in 0..10000 {
		let string_interner: StringInterner<string_interner::DefaultSymbol, Stub, _> =
			StringInterner::with_hasher(hasher.clone());
		threads.push(std::thread::spawn(move || {
			std::thread::sleep(std::time::Duration::from_millis(10));
			let _this_is_memory_unsafe = string_interner;
		}));
	}
	let weak = Rc::downgrade(&hasher.inner);
	std::mem::drop(hasher);
	threads.into_iter().for_each(|t| t.join().unwrap());
	assert!(weak.strong_count() == 0);
}
```
The assert fails though all the `Rc`s are dropped, because it races on separate threads (`Rc`s are not `Send`, and this structure allows me to send them through threads without using `unsafe`). When replacing `Rc`s with `Arc`s, the assert passes.

This is due to this:
https://github.com/Robbepop/string-interner/blob/c2b87fd3b1768fb6d28abe747a4983bfca7b0a0b/src/interner.rs#L278-L292

I do not understand why this `unsafe` was necessary here. It seems the compiler would automatically set `Send`/`!Send` and `Sync`/`!Sync` correctly without these manual implementations.

Overall, I feel like this crate would hugely benefit from dropping its `unsafe` usages (or at least to provide a `Backend` that is fully safe), because these are _actually extremely hard to use_ without endangering memory safety (as I hope I could demonstrate here). 
(As another example of a risk: `InternedStr::as_str` should also be marked `unsafe` in order to underline that it would be a memory bug unless the underlying container has not been dropped, and `PartialEq` should be removed as it's not safe for that same reason. This would ensure that anyone using `InternedStr::as_str` makes sure the container wasn't dropped before calling it.) 

In the general case, I feel I like every `unsafe` use should be benchmarked and show above-noise performance improvements before considering to use it.

**Comment by Robbepop** [comment]
You are right about those impls and I have to admit that I have forgotten to remove them for the recent `0.11` release since they are no longer needed (and actually incorrect) since then.
Also I think you are right that we should get rid of each and every `unsafe` that we can get as long as we can uphold the performance.
For `InternedStr::as_str` I know exactly what you mean. I actually made this function safe while knowing that it should actually be `unsafe` but with hinsight that this is an internal API that is going to be kept internal and also with limited use in the 2 backends.

To make you even more happy: I am planning on implementing another efficient backend based on `strena` that should be entirely safe Rust. :)

**Comment by Ten0** [comment]
> I am planning on implementing another efficient backend based on strena that should be entirely safe Rust. :)

That's great! :) I can't find what `strena` is though. Is it a crate?

> There are currently 5 uses of unsafe remaining in the crate. I am not sure we can tackle all of them but pretty sure most. :)

🎉

> I actually made this function safe while knowing that it should actually be unsafe but with hinsight that this is an internal API that is going to be kept internal and also with limited use in the 2 backends.

I've seen it disappear from the public interface indeed, but I beleive that would still make the crate easier to understand, maintain and contribute to: forcing to write `unsafe` at the exact place where we have to make sure the guarantees are upheld is probably the best way to not forget these guarantees have to be maintained.

**Comment by Robbepop** [comment]
I just released `0.11.1` with your changes incorporated amongst others. ;)

> I've seen it disappear from the public interface indeed, but I beleive that would still make the crate easier to understand, maintain and contribute to: forcing to write unsafe at the exact place where we have to make sure the guarantees are upheld is probably the best way to not forget these guarantees have to be maintained.

Generally I agree, however, for `InternedStr` making it being usable very similar to a `&str` seriously improved the readability of the backends by such a degree that I was in favor of the decision of taking away `unsafe` from the `as_str` operation. Because if it wasn't safe we also couldn't have all the other nice trait impls and they really do their job at a crate private level.
